### DISCUSSION TITLE
Is there a way to mark tooltips and accessibility related strings in Transifex as such?

### DISCUSSION THREAD

**Discussion by StoyanDimitrov** [keepassxreboot/keepassxc#10224] [Q&A] [answered]
As you may have guessed they are kind of different UI paradigm and need to be treated differently. When I wrote "mark" in the title I had in mind things like:
- assign a tag,
- additional context
- anything suitable to tell them apart from "regular" UI strings of buttons, field labels etc.

I don't know what's possible with Transifex.

**Comment by droidmonkey** [comment, marked as answer]
You can add context (disambiguation) to a translated string in qt. See the function signature for `tr`: https://doc.qt.io/qt-5/qobject.html#tr

**Reply by phoerious** [reply]
Most of our tooltips are actually in the XML .ui files, so that context would have to be added there.

**Reply by StoyanDimitrov** [reply]
The name of the .ui file is already set as context for all the strings in it. Unfortunately there are other, non-tooltips strings there so the name of the file alone is not distinct enough.

**Reply by droidmonkey** [reply]
Not the name of the file, but setting context from within Qt Designer. I have done this for most of the accessibility strings. You have to use the arrow in the properties pane to show those options.

**Comment by StoyanDimitrov** [comment]
> You can add context (disambiguation) to a translated string in qt. See the function signature for `tr`: https://doc.qt.io/qt-5/qobject.html#tr

Do you have an idea how that is going to be accessible/visible/presented from the Transifex side?

**Comment by noraj** [comment]
Sometimes it's difficult to pick a translation because of the lack of context. The only context provided on transifex is the class where it is used, e.g. ReportsWidgetBrowserStatistics. Then it's up to you to go on github, find the file where this class is used, then try to understand the context. Wouldn't that be possible to put the github URL pointing to the file and line number like https://github.com/keepassxreboot/keepassxc/blob/0b5ae1775cfb34143ee628694daee69e4765bbc7/src/gui/GuiTools.cpp#L77. Or a list of URL's if used in several contexts.

**Comment by StoyanDimitrov** [comment]
Since I already have been there I'll describe what I found working for me.

I have 4 tabs dedicated to translating KeePassXC:
1. KeePassXC@Transifex (work on the `develop` material - all matching strings will be copied over `stable`) with filter `translated:no`
2. KeePassXC@Transifex with filter `context:(some-context)`
3. KeePassXC@Transifex with filter `text:(some-search)`
4. GitHub https://github.com/search?q=repo%3Akeepassxreboot%2Fkeepassxc "(some-search)"&type=code

I'll describe in detail how I use this setup.
1. From tab 1. above select the first untranslated string (you can select whatever string you are up to) and copy the context of that string: ![tx-1](https://github.com/user-attachments/assets/53794c1d-b500-4aad-8fd9-09edeb7f6815) (sorry, but the fellas @ Transfiex made it difficult to change UI language)
2. Paste the copied context to tab 2. from the list above in the place of `(some-context)` part. Hit Enter. Now all strings from this context are listed. You can look around the list and familiarize, and probably guess where you can find this set of strings in the UI of the application. That's a good thing!
3. Search the repo for the context string with tab 4. from the list above. Note that you'll need a GitHub account for that. Most of the time there will be a `(some-context).ui` file that hold all/most of the strings. From the markup of that file you can obtain information about the nature of the string. Weather it is a tooltip, a label, helper text, window title, dropdown element etc. Knowing the precise context and type of the string return to tab 2. and translate it. Repeat until all untranslated strings within the current filter are done. Proceed to the next step.
4. Use tab 3. to search for inconsistent translations. When several people work on one project (or even you alone over an extended period of time) if they are not agreed upon a dictionary of a king there always will be inconsistencies.
5. KeePassXC should be open and you'll use the gathered information from the steps above to hunt down the desired string. Note that not all strings available in Transifex could be viewed by this or other way because they cover some edge cases of the software lifecycle.

Exercise caution when searching in GitHub because strings from Transifex and GitHub don't always match 1:1. Especially difficulty is finding segmented strings (texts from the UI that are separated into several strings in the repo/Transifex).

That setup gave me assurances where the string I'm looking at is used in the UI so I can translate it knowing the precise context. That lead to speed and quality of translation.

Note all that being said Transifex is inefficient and broken way of handling translations. I'm not even sure they have a way to map string → source code location you request.

**Reply by noraj** [reply]
> KeePassXC@Transifex (work on the `develop` material - all matching strings will be copied over `stable`) with filter `translated:no`

I made translation work on stable branch, completing untranslated sentences. Will this work be overwritten with empty content from the develop branch?

**Reply by StoyanDimitrov** [reply]
No, I think the copying works both way when there is 100% match. The presumption behind translating `develop` is that `develop` receives the new strings (those for the next stable release) and therefore more content to translate. Finishing it almost certainly covers all of `stable` strings.
